File: backend/app/schemas/umatest.py
"""
Pydantic schemas for UMATest module
Phase 1: Test Creation System
"""
from typing import List, Optional, Dict, Any
from datetime import datetime
from pydantic import BaseModel, Field, validator
from uuid import UUID


# Question distribution constants
QUESTION_DISTRIBUTION = {
    "BASIC_INTERMEDIATE_PERCENT": 70,
    "ADVANCED_PERCENT": 20,
    "EXPERT_PERCENT": 10,
    "QUESTIONS_PER_TOPIC": 10
}


class AnswerKey(BaseModel):
    """Answer key for a test question"""
    correct_answer: str
    explanation: str
    evaluation_rubric: str


class TestQuestion(BaseModel):
    """Individual test question"""
    id: str
    question_text: str
    difficulty_level: str
    source_content: str
    answer_key: AnswerKey


class TopicQuestions(BaseModel):
    """Questions for a specific topic"""
    topic_title: str
    source_lecture_id: str
    source_lecture_title: str
    questions: List[TestQuestion]


class GenerationMetadata(BaseModel):
    """Metadata about test generation"""
    generated_at: datetime
    ai_model: str
    distribution: Dict[str, int]


class TestStructure(BaseModel):
    """Complete test structure"""
    total_questions: int
    topics: Dict[str, TopicQuestions]
    generation_metadata: GenerationMetadata


# Hand-built test schemas
class HandBuiltQuestionCreate(BaseModel):
    """Create a hand-built test question"""
    question_text: str = Field(..., min_length=1)
    correct_answer: str = Field(..., min_length=1)
    explanation: str = Field(..., min_length=1)
    evaluation_rubric: str = Field(..., min_length=1)
    difficulty_level: str = Field(..., pattern='^(basic|intermediate|advanced|expert)$')
    points: Optional[int] = Field(10, gt=0)


class HandBuiltQuestionUpdate(BaseModel):
    """Update a hand-built test question"""
    question_text: Optional[str] = Field(None, min_length=1)
    correct_answer: Optional[str] = Field(None, min_length=1)
    explanation: Optional[str] = Field(None, min_length=1)
    evaluation_rubric: Optional[str] = Field(None, min_length=1)
    difficulty_level: Optional[str] = Field(None, pattern='^(basic|intermediate|advanced|expert)$')
    points: Optional[int] = Field(None, gt=0)


class HandBuiltQuestionResponse(BaseModel):
    """Response for a hand-built test question"""
    id: UUID
    test_assignment_id: UUID
    question_text: str
    correct_answer: str
    explanation: str
    evaluation_rubric: str
    difficulty_level: str
    position: int
    points: int
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


# Request schemas
class CreateTestRequest(BaseModel):
    """Request to create a new test"""
    test_title: str = Field(..., min_length=1, max_length=255)
    test_description: Optional[str] = None
    test_type: str = Field('lecture_based', pattern='^(lecture_based|hand_built)$')
    selected_lecture_ids: Optional[List[UUID]] = None
    time_limit_minutes: Optional[int] = Field(None, gt=0)
    attempt_limit: Optional[int] = Field(1, gt=0)
    randomize_questions: Optional[bool] = False
    show_feedback_immediately: Optional[bool] = True
    
    @validator('selected_lecture_ids')
    def validate_lecture_ids(cls, v, values):
        test_type = values.get('test_type', 'lecture_based')
        if test_type == 'lecture_based' and (not v or len(v) == 0):
            raise ValueError("At least one lecture must be selected for lecture-based tests")
        elif test_type == 'hand_built' and v:
            raise ValueError("Hand-built tests should not have selected lectures")
        return v


class UpdateTestRequest(BaseModel):
    """Request to update test settings"""
    test_title: Optional[str] = Field(None, min_length=1, max_length=255)
    test_description: Optional[str] = None
    time_limit_minutes: Optional[int] = Field(None, gt=0)
    attempt_limit: Optional[int] = Field(None, gt=0)
    randomize_questions: Optional[bool] = None
    show_feedback_immediately: Optional[bool] = None
    status: Optional[str] = Field(None, pattern='^(draft|published|archived)$')


class GenerateTestQuestionsRequest(BaseModel):
    """Request to generate test questions"""
    test_assignment_id: UUID
    regenerate: Optional[bool] = False


# Response schemas
class TestAssignmentBase(BaseModel):
    """Base test assignment data"""
    id: UUID
    teacher_id: UUID
    test_title: str
    test_description: Optional[str]
    test_type: str
    selected_lecture_ids: Optional[List[UUID]]
    time_limit_minutes: Optional[int]
    attempt_limit: int
    randomize_questions: bool
    show_feedback_immediately: bool
    status: str
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


class TestAssignmentResponse(TestAssignmentBase):
    """Test assignment response with structure"""
    test_structure: Optional[Dict[str, Any]] = None
    total_questions: Optional[int] = None
    is_archived: Optional[bool] = False
    
    @validator('total_questions', always=True)
    def calculate_total_questions(cls, v, values):
        if 'test_structure' in values and values['test_structure'] and isinstance(values['test_structure'], dict):
            return values['test_structure'].get('total_questions', 0)
        return 0


class LectureInfo(BaseModel):
    """Basic info about a lecture"""
    id: UUID
    title: str
    subject: str
    grade_level: str
    topic_count: int


class TestDetailResponse(TestAssignmentResponse):
    """Detailed test response with lecture info"""
    selected_lectures: Optional[List[LectureInfo]] = None


class TestListResponse(BaseModel):
    """Paginated list of tests"""
    tests: List[TestAssignmentResponse]
    total_count: int
    page: int
    page_size: int


class QuestionGenerationProgress(BaseModel):
    """Progress update for question generation"""
    test_assignment_id: UUID
    status: str
    progress: Dict[str, Any]
    error: Optional[str] = None


class GenerationLogResponse(BaseModel):
    """Generation log details"""
    id: UUID
    test_assignment_id: UUID
    started_at: datetime
    completed_at: Optional[datetime]
    status: str
    error_message: Optional[str]
    total_topics_processed: int
    total_questions_generated: int
    cache_hits: int
    cache_misses: int
    ai_tokens_used: int
    ai_model: Optional[str]
    
    class Config:
        from_attributes = True


# Helper functions
def calculate_question_counts(total_topics: int) -> Dict[str, Any]:
    """Calculate question distribution based on topic count"""
    questions_per_topic = QUESTION_DISTRIBUTION["QUESTIONS_PER_TOPIC"]
    total = total_topics * questions_per_topic
    
    return {
        "total": total,
        "basic_intermediate": round((total * QUESTION_DISTRIBUTION["BASIC_INTERMEDIATE_PERCENT"]) / 100),
        "advanced": round((total * QUESTION_DISTRIBUTION["ADVANCED_PERCENT"]) / 100),
        "expert": round((total * QUESTION_DISTRIBUTION["EXPERT_PERCENT"]) / 100),
        "per_topic": {
            "basic_intermediate": round((questions_per_topic * QUESTION_DISTRIBUTION["BASIC_INTERMEDIATE_PERCENT"]) / 100),
            "advanced": round((questions_per_topic * QUESTION_DISTRIBUTION["ADVANCED_PERCENT"]) / 100),
            "expert": round((questions_per_topic * QUESTION_DISTRIBUTION["EXPERT_PERCENT"]) / 100)
        }
    }

File: types/umatest.ts
// UMATest Module Type Definitions
// Phase 1: Test Creation System

export interface TestAssignment {
  id: string;
  teacher_id: string;
  test_title: string;
  test_description: string | null;
  selected_lecture_ids: string[];
  time_limit_minutes: number | null;
  attempt_limit: number;
  randomize_questions: boolean;
  show_feedback_immediately: boolean;
  test_structure: TestStructure;
  status: 'draft' | 'published' | 'archived';
  created_at: string;
  updated_at: string;
  deleted_at: string | null;
}

export interface TestStructure {
  total_questions: number;
  topics: Record<string, TopicQuestions>;
  generation_metadata: GenerationMetadata;
}

export interface TopicQuestions {
  topic_title: string;
  source_lecture_id: string;
  source_lecture_title: string;
  questions: TestQuestion[];
}

export interface TestQuestion {
  id: string;
  question_text: string;
  difficulty_level: 'basic' | 'intermediate' | 'advanced' | 'expert';
  source_content: string;
  answer_key: AnswerKey;
}

export interface AnswerKey {
  correct_answer: string;
  explanation: string;
  evaluation_rubric: string;
}

export interface GenerationMetadata {
  generated_at: string;
  ai_model: string;
  distribution: {
    basic_intermediate: number;
    advanced: number;
    expert: number;
  };
}

export interface TestQuestionCache {
  id: string;
  lecture_id: string;
  topic_id: string;
  difficulty_level: 'basic' | 'intermediate' | 'advanced' | 'expert';
  content_hash: string;
  questions: CachedQuestion[];
  ai_model: string;
  generation_timestamp: string;
}

export interface CachedQuestion {
  question_text: string;
  answer_key: AnswerKey;
  source_excerpt: string;
}

export interface TestGenerationLog {
  id: string;
  test_assignment_id: string;
  started_at: string;
  completed_at: string | null;
  status: 'processing' | 'completed' | 'failed';
  error_message: string | null;
  total_topics_processed: number;
  total_questions_generated: number;
  cache_hits: number;
  cache_misses: number;
  ai_tokens_used: number;
  ai_model: string | null;
}

// API Request/Response Types

export interface CreateTestRequest {
  test_title: string;
  test_description?: string;
  selected_lecture_ids: string[];
  time_limit_minutes?: number;
  attempt_limit?: number;
  randomize_questions?: boolean;
  show_feedback_immediately?: boolean;
}

export interface UpdateTestRequest {
  test_title?: string;
  test_description?: string;
  time_limit_minutes?: number | null;
  attempt_limit?: number;
  randomize_questions?: boolean;
  show_feedback_immediately?: boolean;
  status?: 'draft' | 'published' | 'archived';
}

export interface GenerateTestQuestionsRequest {
  test_assignment_id: string;
  regenerate?: boolean; // Force regeneration even if questions exist
}

export interface TestListResponse {
  tests: TestAssignment[];
  total_count: number;
  page: number;
  page_size: number;
}

export interface TestDetailResponse extends TestAssignment {
  selected_lectures?: LectureInfo[]; // Populated lecture information
}

export interface LectureInfo {
  id: string;
  title: string;
  subject: string;
  grade_level: string;
  topic_count: number;
}

export interface QuestionGenerationProgress {
  test_assignment_id: string;
  status: 'processing' | 'completed' | 'failed';
  progress: {
    current_topic: number;
    total_topics: number;
    current_topic_name?: string;
  };
  error?: string;
}

// Distribution configuration for question generation
export const QUESTION_DISTRIBUTION = {
  BASIC_INTERMEDIATE_PERCENT: 70,
  ADVANCED_PERCENT: 20,
  EXPERT_PERCENT: 10,
  QUESTIONS_PER_TOPIC: 10
} as const;

// Helper function to calculate question counts
export function calculateQuestionCounts(totalTopics: number) {
  const questionsPerTopic = QUESTION_DISTRIBUTION.QUESTIONS_PER_TOPIC;
  const total = totalTopics * questionsPerTopic;
  
  return {
    total,
    basicIntermediate: Math.round((total * QUESTION_DISTRIBUTION.BASIC_INTERMEDIATE_PERCENT) / 100),
    advanced: Math.round((total * QUESTION_DISTRIBUTION.ADVANCED_PERCENT) / 100),
    expert: Math.round((total * QUESTION_DISTRIBUTION.EXPERT_PERCENT) / 100),
    perTopic: {
      basicIntermediate: Math.round((questionsPerTopic * QUESTION_DISTRIBUTION.BASIC_INTERMEDIATE_PERCENT) / 100),
      advanced: Math.round((questionsPerTopic * QUESTION_DISTRIBUTION.ADVANCED_PERCENT) / 100),
      expert: Math.round((questionsPerTopic * QUESTION_DISTRIBUTION.EXPERT_PERCENT) / 100)
    }
  };
}